ops/testing.py

```diff
@@ -2425,66 +2537,71 @@ def push(
             raise pebble.PathError(
                 'generic-file-error',
                 f'permissions not within 0o000 to 0o777: {permissions:#o}')
+        self.__check_absolute_path(path)
+        file_path = self._root / path[1:]
+        if make_dirs and not file_path.parent.exists():
+            self.make_dir(
+                os.path.split(path)[0],
+                make_parents=True,
+                permissions=None,
+                user_id=user_id,
+                user=user,
+                group_id=group_id,
+                group=group)
+        permissions = permissions if permissions is not None else 0o644
         try:
-            self._fs.create_file(
-                path, source, encoding=encoding, make_dirs=make_dirs, permissions=permissions,
-                user_id=user_id, user=user, group_id=group_id, group=group)
+            if isinstance(source, str):
+                content = source.encode(encoding=encoding)
+            elif isinstance(source, bytes):
+                content = source
+            else:
+                source_read: Union[str, bytes] = source.read()
+                content = (
+                    source_read
+                    if isinstance(source_read, bytes)
+                    else source_read.encode(encoding=encoding)
+                )
+            file_path.write_bytes(content)
```

Comment:
In the `str` and `bytes` cases we already have the content in memory, but I think we should avoid reading the whole thing into memory if it's a file/buffer. Can we instead use `copyfileobj`, which only reads in 64K chunks, and do something like this (untested):

```python
if isinstance(source, str):
    file_path.write_text(source, encoding=encoding)
elif isinstance(source, bytes):
    file_path.write_bytes(source)
else:
    with file_path.open('w' if encoding is None else 'wb', encoding=encoding) as f:
        shutil.copyfileobj(source, f)
```

Reply:
It's better, updated, thanks!